File: .Test-Demo/PhotoAssessmentUsageDemo/PhotoAssessmentUsageDemo/ViewController.m
//
//  ViewController.m
//  PhotoAssessmentUsageDemo
//
//  Created by 杨萧玉 on 2018/11/20.
//  Copyright © 2018 杨萧玉. All rights reserved.
//

#import "ViewController.h"
#import "PhotoScanner.h"
#import <Photos/Photos.h>
#import "PhotoAssessmentHelper.h"
#import "TZImagePickerController/TZImagePickerController.h"

@interface ViewController ()

@property (nonatomic, strong) PhotoAssessmentHelper *helper;

@property (weak, nonatomic) IBOutlet UIButton *button;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
//    self.helper = [PhotoAssessmentHelper new];
//
//    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
//        NSDate *start = [NSDate date];
//        NSArray<PHAsset *> *assets = [PhotoScanner scanPhotosWithLimit:500];
//        dispatch_group_t group = dispatch_group_create();
//        dispatch_queue_t queue = dispatch_queue_create("com.photoassessment.demo", NULL);
//        NSMutableArray<PhotoAssessmentResult *> *results = [NSMutableArray array];
//        PHImageRequestOptions *options = [PHImageRequestOptions new];
//        options.resizeMode  = PHImageRequestOptionsResizeModeFast;
//        options.deliveryMode = PHImageRequestOptionsDeliveryModeHighQualityFormat;
//        options.networkAccessAllowed = NO;
//        for (PHAsset *asset in assets) {
//            dispatch_group_enter(group);
//            // targetSize 的调整会影响计算总耗时，但并不是 targetSize 越小耗时越小，好神奇。大概试了下，500最合适
//            [[PHImageManager defaultManager] requestImageForAsset:asset
//                                                       targetSize:CGSizeMake(240, 240)
//                                                      contentMode:PHImageContentModeAspectFit
//                                                          options:options
//                                                    resultHandler:^(UIImage *image, NSDictionary *info) {
//                                                        if (image.CGImage) {
//                                                            [self.helper requestMPSAssessmentScoreFor:image.CGImage completionHandler:^(PhotoAssessmentResult * _Nonnull result) {
//                                                                dispatch_async(queue, ^{
//                                                                    [results addObject:result];
//                                                                    dispatch_group_leave(group);
//                                                                    NSLog(@"%@", result);
//                                                                });
//                                                            }];
//                                                        }
//                                                        else {
//                                                            dispatch_group_leave(group);
//                                                        }
//                                                    }];
//        }
//        dispatch_group_notify(group, queue, ^{
//            NSTimeInterval duration = -[start timeIntervalSinceNow];
//            NSLog(@"total complete %f", duration);
//            UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Finished" message:[NSString stringWithFormat:@"Duration:%f", duration] preferredStyle:UIAlertControllerStyleAlert];
//            [self presentViewController:alert animated:YES completion:nil];
//        });
//    });
}

- (IBAction)clickButton:(id)sender {
    TZImagePickerController *imagePickerVc = [[TZImagePickerController alloc] init];
    imagePickerVc.allowPickingVideo = NO;
    imagePickerVc.allowPickingGif = NO;
    [self presentViewController:imagePickerVc animated:YES completion:nil];
}

@end

File: PhotoAssessment-Sample/PhotoAssessment/ViewController.swift
//
//  ViewController.swift
//  PhotoAssessment
//
//  Created by 杨萧玉 on 2018/11/9.
//  Copyright © 2018 杨萧玉. All rights reserved.
//

import UIKit
import Photos

class ViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var cameraButton: UIBarButtonItem!
    @IBOutlet weak var assessmentLabel: UILabel!
    @IBOutlet weak var detailLabel: UILabel!
    
    @available (iOS 11.0, *)
    lazy var helper: PhotoAssessmentHelper = {
       return PhotoAssessmentHelper()
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let status = PHPhotoLibrary.authorizationStatus()
        if status == .notDetermined {
            PHPhotoLibrary.requestAuthorization { (status) in
                
            }
        }
    }
    
    @IBAction func takePicture(_ sender: UIBarButtonItem) {
        guard UIImagePickerController.isSourceTypeAvailable(.camera) else {
            presentPhotoPicker(sourceType: .photoLibrary)
            return
        }
        
        let photoSourcePicker = UIAlertController()
        let choosePhoto = UIAlertAction(title: "Choose Photo", style: .default) { [unowned self] _ in
            self.presentPhotoPicker(sourceType: .photoLibrary)
        }
        
        photoSourcePicker.addAction(choosePhoto)
        photoSourcePicker.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        
        present(photoSourcePicker, animated: true)
    }
    
    func presentPhotoPicker(sourceType: UIImagePickerController.SourceType) {
        let picker = UIImagePickerController()
        picker.delegate = self
        picker.sourceType = sourceType
        present(picker, animated: true)
    }
    
}

extension ViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picker.dismiss(animated: true)
        let image = info[.originalImage] as! UIImage
        imageView.image = image
        assessmentLabel.text = "Processing..."
        detailLabel.text = ""
        if let asset = info[.phAsset] as? PHAsset {
            DispatchQueue.global().async {
                guard asset.mediaType == .image else {
                    return
                }
                let options = PHImageRequestOptions()
                options.deliveryMode = .highQualityFormat
                options.resizeMode = .fast
                PHImageManager.default().requestImage(for: asset, targetSize: CGSize(width: 500, height: 500), contentMode: .aspectFit, options: options, resultHandler: { (image, info) in
                    guard let downsampleImage = image?.cgImage else {
                        return
                    }
                    self.helper.requestSubjectiveAssessment(for: downsampleImage, completionHandler: { (score) in
                        DispatchQueue.main.async {
                            self.assessmentLabel.text = String(format: "Assessment Score:%0.5f", score)
                        }
                    })
                    self.helper.requestObjectiveAssessment(for: downsampleImage, downsampleDimension: 50, completionHandler: { (result) in
                        DispatchQueue.main.async {
                            self.detailLabel.text = result.description
                        }
                    })
//                    if let cgImage = image?.cgImage {
//                        let imagePixels = cgImage.rgbPixels()
//                        var date = Date()
//                        let fingerprint = Utils.fingerprint(ofImagePixels: imagePixels, width: cgImage.width, height: cgImage.height)
//                        print("cpu fingerprint cost: \(-date.timeIntervalSinceNow)")
//                        date = Date()
//                        let hsb = Utils.meanHSB(ofImagePixels: imagePixels, width: cgImage.width, height: cgImage.height)
//                        print("cpu meanSaturation cost: \(-date.timeIntervalSinceNow)")
//                        let mpsProcessor = PhotoMPSProcessor()
//                        date = Date()
//                        mpsProcessor.fingerprint(ofImagePixels: imagePixels, width: cgImage.width, height: cgImage.height, completionHandler: { (result) in
//                            print("gpu fingerprint cost: \(-date.timeIntervalSinceNow)")
//                            print("equal result:\(fingerprint == result)")
//                        })
//                        mpsProcessor.meanSaturation(ofImagePixels: imagePixels, width: cgImage.width, height: cgImage.height, completionHandler: { (result) in
//                            print("gpu meanSaturation cost: \(-date.timeIntervalSinceNow)")
//                            print("equal result:\(Float(hsb.saturation) - result < 0.0001)")
//                        })
//                    }
                })
            }
        }
    }
    
}

